be/src/exec/range_router.cpp

```diff
@@ -0,0 +1,239 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/range_router.h"
+
+#include <algorithm>
+#include <boost/iterator/counting_iterator.hpp>
+
+#include "column/column_helper.h"
+#include "column/datum_convert.h"
+#include "column/schema.h"
+#include "common/statusor.h"
+#include "fmt/format.h"
+#include "runtime/descriptors.h"
+#include "storage/types.h"
+
+namespace starrocks {
+
+Status RangeRouter::init(const std::vector<TTabletRange>& tablet_ranges, size_t num_columns) {
+    const size_t num_ranges = tablet_ranges.size();
+    if (num_ranges == 0) {
+        return Status::InternalError("no tablet ranges for RangeRouter init");
+    }
+    if (!_upper_boundaries.empty()) {
+        DCHECK((_upper_boundaries[0] == nullptr) ||
+               (_upper_boundaries[0]->size() + 1 == num_ranges));
+        return Status::OK();
+    }
+
+    RETURN_IF_ERROR(_validate_range(tablet_ranges, num_columns));
+
+    _upper_boundaries.resize(num_columns);
+    for (size_t i = 0; i < num_ranges - 1; ++i) {
+        _upper_boundaries_slice.emplace_back(&_upper_boundaries, i);
```

Comment:
should check the ranges is ordered if use a vector

Reply:
`_validate_range` can ensure that